### DISCUSSION TITLE
Lesson 11: MOOD NFT with OpenZeppelin  v5.0.0

### DISCUSSION THREAD

**Discussion by AlSirang** [Cyfrin/foundry-full-course-cu#1075] [General]
Hello everyone,

I was using OpenZeppelin Contracts v5.0.0 (token/ERC721/ERC721.sol) to create `MoodNFT` and encountered an error while calling the `_isApprovedOrOwner` function. It turned out that this function has been removed. If you are also using v5, please use the following approach.

```diff

  function flipMood(uint256 tokenId) public {
--       if (!_isApprovedOrOwner(msg.sender, tokenId)) {
--          revert MoodNft__CantFlipMoodIfNotOwner();
--       }

++      address from = _ownerOf(tokenId);
++      _checkAuthorized(from, msg.sender, tokenId);

        if (s_tokenIdToState[tokenId] == NFTState.HAPPY) {
            s_tokenIdToState[tokenId] = NFTState.SAD;
        } else {
            s_tokenIdToState[tokenId] = NFTState.HAPPY;
        }
    }


```

Another function that has been removed is `_exists`, and its replacement is `_requireOwned`.

```diff

  function tokenURI(
        uint256 tokenId
    ) public view virtual override returns (string memory) {
--        if (!_exists(tokenId)) {
--            revert ERC721Metadata__URI_QueryFor_NonExistentToken();
--       }
++       _requireOwned(tokenId);

       // rest of code
        
    }


```

**Comment by AbhiSharma1210** [comment]
I suspected that `_ownerOf(tokenId) = msg.sender` 
Also, in the openzeppelin changelog. [Here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/CHANGELOG.md)
They mentioned the following 

> In this logic of removing hidden SLOADs, the _isApprovedOrOwner function was removed in favor of a new _isAuthorized function. Overrides that used to target the _isApprovedOrOwner should now be performed on the _isAuthorized function.

So `_isAuthorized` can be used like 
`
if (!_isAuthorized(AddressA, AddressB, tokenId)) {
            revert ColorNft__CantFlipColorIfNotOwner();
        }
`
is it correct ?
